File: EmuSen/Cores/Nintendo/Mars - N64/Memory/IsViewer.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace EmuSen.Cores.Nintendo.Mars.Memory
{
    // The debug text port the hardware corpus prints through - see Mars_Memory.md §4.
    public sealed class IsViewer
    {
        public const uint LengthRegisterOffset = 0x14;
        public const uint BufferOffset = 0x20;

        // Readable RAM, because the corpus detects the port by reading back what it wrote - see §4.1.
        private readonly byte[] _memory = new byte[MemoryMap.IsViewerSize];

        // The harness's transcript of the port, not the port - see Mars_SaveStates.md §2.
        [EmuSen.Common.SkipInState] private readonly List<byte> _text = new();

        public IReadOnlyList<byte> Captured => _text;

        public string Text => Encoding.UTF8.GetString(_text.ToArray());

        public uint Read32(uint offset) =>
            (uint)((_memory[offset] << 24) | (_memory[offset + 1] << 16) | (_memory[offset + 2] << 8) | _memory[offset + 3]);

        public void Write32(uint offset, uint value)
        {
            _memory[offset] = (byte)(value >> 24);
            _memory[offset + 1] = (byte)(value >> 16);
            _memory[offset + 2] = (byte)(value >> 8);
            _memory[offset + 3] = (byte)value;

            if (offset == LengthRegisterOffset && value != 0) Flush(value);
        }

        public byte Read8(uint offset) => _memory[offset];

        public void Write8(uint offset, byte value) => _memory[offset] = value;

        // A length write means "emit this many bytes from the buffer", and the register keeps its value.
        private void Flush(uint length)
        {
            uint available = MemoryMap.IsViewerSize - BufferOffset;
            uint count = Math.Min(length, available);

            for (uint i = 0; i < count; i++) _text.Add(_memory[BufferOffset + i]);
        }

        public void Clear() => _text.Clear();
    }
}

File: EmuSen/Cores/Nintendo/MarsRT - N64/src/isviewer.rs
//! The IS-Viewer, the debug text port the hardware corpus prints through: the C# `IsViewer`. See Mars_Memory.md §4.

use crate::Skip;
use crate::state::{State, StateReader, StateResult, StateWriter, boxed};

pub const LENGTH_REGISTER_OFFSET: u32 = 0x14;
pub const BUFFER_OFFSET: u32 = 0x20;
const SIZE: u32 = 0x1000;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct IsViewer {
    pub memory: Box<[u8; 4096]>,
    /// `_text`: the harness's transcript of the port, not the port.
    pub text: Skip<Vec<u8>>,
}

impl Default for IsViewer {
    fn default() -> Self {
        IsViewer { memory: boxed(0), text: Skip(Vec::new()) }
    }
}

impl State for IsViewer {
    fn write_state(&self, w: &mut StateWriter) {
        w.bytes("_memory", &self.memory[..]);
    }

    fn read_state(&mut self, r: &mut StateReader) -> StateResult {
        r.bytes(&mut self.memory[..])?; // _memory
        Ok(())
    }
}

impl IsViewer {
    pub fn read32(&self, offset: u32) -> u32 {
        let at = offset as usize;
        u32::from_be_bytes([self.memory[at], self.memory[at + 1], self.memory[at + 2], self.memory[at + 3]])
    }

    /// A length write emits that many bytes from the buffer, and the register keeps its value.
    pub fn write32(&mut self, offset: u32, value: u32) {
        let at = offset as usize;
        self.memory[at..at + 4].copy_from_slice(&value.to_be_bytes());
        if offset == LENGTH_REGISTER_OFFSET && value != 0 {
            let count = value.min(SIZE - BUFFER_OFFSET) as usize;
            let from = BUFFER_OFFSET as usize;
            self.text.extend_from_slice(&self.memory[from..from + count]);
        }
    }
}
